Add first draft of code review guidelines

hultonha opened the pull request:
## What does this PR do?

This PR aims to start the process of having a document we as O3DE contributors can all refer to (and improve and extend over time) to make PRs better for code authors and reviewers alike. We want to ensure new and existing contributors feel welcome and included when creating a PR and that they are setup for success.

This is a follow-on from this RFC - https://github.com/o3de/community/issues/140

lmbr-pip commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+# O3DE Code Review Guidelines
+
+This document contains a set of guidelines for code reviewers and change authors when contributing to Open 3D Engine (O3DE). The intention is to help pull requests (PRs) go smoothly and help the community share best practices/lessons learned when it comes to changes and reviews.
```
```suggestion
This document contains a set of guidelines for code reviewers and Pull Request (PR) authors when contributing to Open 3D Engine (O3DE). The intention is to help PRs go smoothly and help the community share best practices and lessons learned when it comes to changes and reviews.
```

lmbr-pip commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+- [O3DE C++ Coding Standards](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#o3de-c-coding-standards-)
+- [O3DE API Ref Guidelines](https://github.com/o3de/sig-core/blob/main/governance/API-Ref-Guidelines-Update.md)
+
+## Change Author Guidelines
```
```suggestion
## Pull Request Creator Guidelines
```

hultonha replied:
I quite like the term 'Change Author' and this is how Google refers to them in their code review guidelines. It's definitely not a hill I'm willing to die on but I just want to check do we think 'Pull Request Creator' is a better alternative? 🤔 Technically you could create the PR without authoring it for example 😝 I could add a little section at the top to define the term 'Change Author' more clearly perhaps?

AMZN-stankowi replied:
I can think of a few cases where the person posting the pull request is not the change author. The most obvious is if a team was working together on something bigger in a branch, and the pull request is to bring that branch work to Development.

Another is if we have a pull request from a contributor stall out, sometimes it makes sense for someone else to pick up the pull request to bring it to the finish line, although in that case "Pull Request Creator" isn't fully accurate, maybe "Pull Request Maintainer"?

hultonha replied:
Hmm I still do think 'Change Author' can cover that, but ideally a definition of these more unusual cases could be included for clarity, I'll update this shortly

lmbr-pip commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+
+The best way to catch simple mistakes and small oversights is to read through your own PR in the GitHub UI before hitting the publish button. Viewing your own code through this slightly different lens often uncovers a lot more issues than you'd expect and can help save other reviewers time.
+
+### PRs are complete
```
```suggestion
### Ensure PRs are ready for review
```

lmbr-pip commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+# O3DE Code Review Guidelines
```
You have inconsistent casing in the section headings. Decide on Camel Capitalization, Sentence Capitalization etc.

hultonha replied:
I can fix this, good spot

hultonha replied:
(should now be fixed to follow 'Sentence capitalization')

lmbr-pip commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+
+### A picture is worth a thousand words
+
+If your change results in any kind of visual difference, to help reviewers quickly get context on the PR, favor including screenshots and/or videos/gifs of the change in action (before/after is also very useful). This helps frame the change and can make reviewers lives much easier to see what the effect of the change is. This goes for test output too.
```
Please emphasize that adding UX screenshots is essential for reviewing these type  of changes.

lmbr-pip commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+
+### Nudges are okay
+
+If your PR has been waiting a while and no one has looked at it, it's okay to start using gentle nudges to ask for feedback. This most likely will be on Discord but it's possible to also tag people in the PR comments and politely ask them to take a look. Code owners will be automatically notified but sometimes things can be missed.
```
Do folks understand what you mean by nudge here? Consider adding links to GitHubs [@mention system](https://github.blog/2011-03-23-mention-somebody-they-re-notified/) for mentioning folks. May also want to add something about reaching out to the relevant SIG on Discord as well.

hultonha replied:
I do mention 'tagging' people on the PR and Discord but I could also put the @mention system in brackets too to be super clear

hultonha replied:
(now added)

lmbr-pip commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+
+If your PR has been waiting a while and no one has looked at it, it's okay to start using gentle nudges to ask for feedback. This most likely will be on Discord but it's possible to also tag people in the PR comments and politely ask them to take a look. Code owners will be automatically notified but sometimes things can be missed.
+
+> It's inevitable people get busy or will be away too, it happens. When faced with this, try and switch to something else or look for another code owner with domain knowledge who can approve the PR.
```
Maybe this is where we add a note like 
```suggestion
> It's inevitable people get busy or will be away too, it happens. When faced with this, try and switch to something else or look for another code owner with domain knowledge who can approve the PR. You can ask for reviews in the relevant SIG channels on Discord.
```

hultonha replied:
This is a good callout, I've updated the wording slightly but the sentiment remains the same.

lmbr-pip commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+
+After reviewing a PR, keep an eye on when the change author has responded and do your best to reply to their responses promptly (ideally within a 24 hour window, though this isn't always possible). If you are delayed and don't have time to respond fully, leave a comment on the PR that you're busy and will get back to it as soon as you can.
+
+### Approve if you can
```
Perhaps add a note about how we don't want Perfect to get in the way of progress. Consider moving some requests for change to a followup GitHub issues.

We may also want to link to the case where folks have not added unit tests. For example, reviewers may want to consider raising with the SIG to see if someone can take on adding unit tests in a subsequent PR. We should acknowledge that not all contributors have infinite time to cover automated testing needs.

hultonha replied:
I've added a reference to the aphorism 'perfect is the enemy of good', good suggestion. To be honest I'd rather not encourage PRs without unit tests (or some kind of test coverage) that is then done later (it often isn't). Depending on the situation either we can afford _not_ to have a test, or you can't make that change without a test... maybe we need a section for that? But I'd rather leave it for now (in the doc we do already mention it being okay to move feedback/changes to a follow-up PR)

lmbr-pip reviewed (commented):
Tom - this is awesome. Its a really wonderful document.

My comments are mostly about how could we improve this, but I don't want to let perfect stand in the way of progress.

AMZN-stankowi commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+
+### Don't be late to the party
+
+If a PR already has two approvals (from at least one code owner) be conscious that additional feedback will interrupt the flow of a PR. Feel free to review and if a serious issue is spotted do feel empowered to take the necessary steps (request changes) but try to avoid lots of additional comments that could be addressed in a follow-up PR.
```
I think late feedback is good, I'd rather get it than someone hold back on good feedback. I think instead this should be phrased along the lines of being accepting of code going in and having the changes responded to in a later follow up pull request. I even don't mind and would rather get feedback on an already merged pull request than miss out on that feedback entirely.

Otherwise, we don't really have a good mechanism for annotating suggested changes outside of pull requests.

hultonha replied:
I've thought about this and do agree, I've just seen some awkward situations where a PR is blocked (request changes) for minor feedback even after have multiple approvals. This guidance is to try and avoid that situation as it can be incredibly demoralizing for contributors. I'll try and be a little more accepting of these changes saying it's okay to comment, but unless very serious, refrain from adding blocking changes and be comfortable about moving the updates to a fast-follow PR

AMZN-stankowi commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+
+### Keep feedback focused
+
+Ensure feedback is targeted at the changed lines of code and not surrounding code. It isn't the change authors responsibility to fix a problem a reviewer sees that happens to be in that area of code. If something is spotted, make an issue, and follow-up with a separate PR.
```
I'd rather have the reviewer first ask if the pull request author is willing to make a change and be absolutely willing to take a "Sorry out of scope for this pull request" reply. Often it's much faster to make a small change to existing code, than the effort to write a follow up ticket, spin up a new branch to do that change, generate a PR for that change, and run automated review.

hultonha replied:
Again this is true it's just a question of scope, and how much is asked. I think again I can clarify this okay if it's small, or if it can be moved to a separate PR to happen right after. It isn't cool though to keep piling on suggested changes in the surrounding code when the contributor has tasks to complete and things to do outside of that focus area

AMZN-stankowi commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+
+_In general, reviewers should favor approving a [PR] once it is in a state where it definitely improves the overall code health of the system being worked on, even if the [PR] isn’t perfect._
+
+It's inevitable all changes will not get everything right, but if what is there improves the software while keeping it maintainable and efficient, a reviewer should feel empowered to approve.
```
this might be a good place to remind people we can always make follow up tickets. I find it can be useful to call something out in a pull request "This could use an automated test" but with a note "and I've created a ticket to backfill that automated test, we don't need to hold up this change"

AMZN-stankowi commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+
+## Contributors, Reviewers and Maintainers
+
+Once a PR is approved by two or more reviewers, an AR run can commence. Currently only [maintainers](https://github.com/orgs/o3de/teams/maintainers) can initiate builds so please feel free to ask for someone with access to start a build. The list of PRs can be found here - [Change Requests](https://jenkins.build.o3de.org/job/O3DE/view/change-requests/). Search for the PR number to review the curr ...
```
We probably want a section to call out the details of this process a bit more clearly. There's a gotcha in our automated review process that requires pull request authors to really be on top of things: The pull request just runs on the source branch, not the merge, so pull request authors need to 1) Merge from the target branch as closely to the time they run automated review as possible, 2) Merge in the pull request as closely to the time the automated review finishes as possible, and 3) Keep an eye on the target branch build after merging, because that Jenkins job does some different things than automated review and can fail when AR passed.

hultonha replied:
I'm maybe okay with this but I'm not sure this is the right document for going into all that detail. I think this should be visible for maintainers/contributors but this is really trying to focus on PR feedback (giving and receiving) as much as possible. I encourage you to create a separate doc to enumerate some of the process steps you've mentioned.

hultonha commented:
Thanks for the feedback so far, I've responded to and made several updates suggested by @lmbr-pip, I'll do my best to take a look at your comments @AMZN-stankowi later this week. Thanks for taking a look! Much appreciated 👍

hultonha commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+
+## Useful Related Material
+
+- [O3DE C++ Coding Standards](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#o3de-c-coding-standards-)
```
Are those public yet? I don't see them listed here - https://github.com/o3de/sig-core/tree/main/governance. If you can please share the link with me I'd be happy to add it here. Maybe someone from @o3de/sig-testing knows where they are?

AMZN-stankowi replied:
I'm not sure, I just assumed we would have them somewhere. Maybe add a note that Python standards are coming later? Otherwise I guess we can wait until someone else adds them?

hultonha replied:
I'd love if someone would add them 😅 But yeah afaik know nothing exists right now, maybe something for @Kadino to investigate? 🙂

hultonha commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,111 @@
+- [O3DE C++ Coding Standards](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#o3de-c-coding-standards-)
+- [O3DE API Ref Guidelines](https://github.com/o3de/sig-core/blob/main/governance/API-Ref-Guidelines-Update.md)
+
+## Change Author Guidelines
```
That's a good callout, I can include a mention of that somewhere

lmbr-pip commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,117 @@
+
+### Keep it short
+
+This of course isn't always possible, there will be exceptions, but to get a PR approved and merged in a reasonable time-frame, try to keep changes as contained and atomic as possible. Small incremental progress is much safer and simpler than a single large change that contains a many modifed lines and files.
```
nit: maybe change "atomic" to be in simpler english. maybe "as focused and self-contained as possible".

lmbr-pip commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,117 @@
+
+### A picture is worth a thousand words
+
+If your change results in any kind of visual difference, to help reviewers quickly get context on the PR, please attach screenshots and/or videos/gifs of the change in action (including before/after is also very useful). This helps frame the change and can make reviewers lives much easier to see what the effect of the change is. This goes for test output too.
```
nit: Maybe add the example of adding screenshots of script canvas graphs as an example, the source assets are almost impossible to review but a visual overview can be very illuminating.

hultonha replied:
Good suggestion, I'll include a mention of this at the end along with the bit about test output.

lmbr-pip commented on guides/o3de-code-review-guidelines.md:
```diff
@@ -0,0 +1,117 @@
+
+### If you can't approve, take things offline
+
+If there is a point of discussion on a PR that cannot be quickly resolved, consider reaching out directly to the change author to talk it over (either over Discord, a separate GitHub discussion or even an RFC - sometimes a voice call is even better if possible). In our experience this leads to better outcomes for both parties and avoids potential misunderstandings. Leaving a brief update on the P ...
```
nit: Maybe, if you need help from other contributors raise it in the relevant SIG channel or public meetings.

lmbr-pip reviewed (approved):
LG as is, some minor thoughts but can publish as is

hultonha commented:
I made my best attempt and addressing the feedback from @lmbr-pip and have now merged the first version. I'm happy for others to now take on making any tweaks/adjustments/improvements going forward. Thanks all for your time!
